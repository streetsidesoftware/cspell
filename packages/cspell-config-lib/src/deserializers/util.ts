export function detectIndent(content: string): string {
    const m = content.match(/^[ \t]+/m);
    return (m && m[0]) || '  ';
}

export function detectIndentAsNum(content: string): number {
    const indent = detectIndent(content).replace(/\t/g, '    ').replace(/[^ ]/g, '');
    return indent.length;
}
