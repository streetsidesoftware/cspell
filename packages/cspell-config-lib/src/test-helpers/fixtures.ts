import * as path from 'path';

const fixturesDir = path.join(__dirname, '../../fixtures');

export function fixtures(pathOfFixture = ''): string {
    return path.resolve(fixturesDir, pathOfFixture);
}
