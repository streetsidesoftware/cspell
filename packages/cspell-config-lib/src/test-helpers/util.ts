import { promises as fs } from 'fs';
import * as path from 'path';

export function json(obj: unknown, indent: string | number = 2): string {
    return JSON.stringify(obj, null, indent) + '\n';
}

export function tempPath(file: string): string {
    const testState = expect.getState();
    return path.join(__dirname, '../../temp', testState.currentTestName || 'test', file);
}

export async function createPathForFile(file: string): Promise<void> {
    await fs.mkdir(path.dirname(file), { recursive: true });
}

export async function copyFile(fromFile: string, toFile: string): Promise<void> {
    await createPathForFile(toFile);
    return fs.copyFile(fromFile, toFile);
}
