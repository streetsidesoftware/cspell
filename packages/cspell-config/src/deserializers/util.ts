export function detectIndent(content: string): string {
    const m = content.match(/^[ \t]+/m);
    return (m && m[0]) || '  ';
}
