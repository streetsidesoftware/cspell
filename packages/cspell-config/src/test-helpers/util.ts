import * as path from 'path';
import { promises as fs } from 'fs';

export function json(obj: unknown, indent: string | number = 2): string {
    return JSON.stringify(obj, null, indent) + '\n';
}

export function tempPath(file: string): string {
    const testState = expect.getState();
    return path.join(__dirname, '../../temp', testState.currentTestName, file);
}

export async function createPathForFile(file: string): Promise<void> {
    await fs.mkdir(path.dirname(file), { recursive: true });
}

export async function copyFile(fromFile: string, toFile: string): Promise<void> {
    await createPathForFile(toFile);
    return fs.copyFile(fromFile, toFile);
}
