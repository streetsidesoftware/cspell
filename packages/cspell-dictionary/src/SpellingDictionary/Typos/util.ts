import { opConcatMap, opFilter, pipe } from '@cspell/cspell-pipe/sync';
import type { TypoEntry, TyposDef, TyposDefKey, TyposDefValue, TypoValueWithSuggestions } from './typos';

/**
 * Append an entry to a TyposDef.
 * @param def - modified in place
 * @param entry- entry to add.
 * @returns def
 */
export function appendToDef(def: TyposDef, entry: TypoEntry | undefined): TyposDef {
    if (!entry) return def;
    if (typeof entry === 'string') {
        def[entry] = false;
        return def;
    }
    if (Array.isArray(entry)) {
        const [key, ...sugs] = entry.map((s) => s.trim());
        if (!key) return def;
        const s = sugs.map((s) => s.trim()).filter((s) => !!s);
        def[key] = !s.length ? false : s.length === 1 ? s[0] : s;
        return def;
    }

    Object.assign(def, entry);
    return def;
}

export function createTyposDef(entries?: Iterable<[TyposDefKey, TyposDefValue]>): TyposDef {
    const def: TyposDef = Object.create(null);

    if (!entries) return def;

    for (const [key, value] of entries) {
        def[key] = isDefined(value) ? value : false;
    }

    return def;
}

/**
 * Extract all suggestions.
 * @param typosDef - the def
 * @returns the set of suggestions.
 */
export function extractAllSuggestions(typosDef: TyposDef): Set<string> {
    const allSugs = pipe(
        Object.values(typosDef),
        opFilter(hasSuggestions),
        opConcatMap((v) => (Array.isArray(v) ? v : [v]))
    );
    return new Set(allSugs);
}

/**
 * Extract all words that have been explicitly ignore because they contains the `ignorePrefix`.
 * @param typosDef - the def
 * @param ignorePrefix - prefix
 * @returns set of ignored words with the prefix removed.
 */
export function extractIgnoreValues(typosDef: TyposDef, ignorePrefix: string): Set<string> {
    const pfxLen = ignorePrefix.length;
    return new Set(
        Object.keys(typosDef)
            .filter((k) => k.startsWith(ignorePrefix))
            .map((k) => k.slice(pfxLen))
    );
}

function isDefined<T>(v: T | undefined | null): v is T {
    return v !== undefined && v !== null;
}

function isString(v: unknown): v is string {
    return typeof v === 'string';
}

function isArray<T>(v: T[] | unknown): v is T[] {
    return Array.isArray(v);
}

function hasSuggestions(v: TyposDefValue): v is TypoValueWithSuggestions {
    return isString(v) || isArray(v);
}
