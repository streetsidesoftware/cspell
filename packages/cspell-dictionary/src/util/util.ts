export function isDefined<T>(v: T | undefined): v is T {
    return v !== undefined;
}
