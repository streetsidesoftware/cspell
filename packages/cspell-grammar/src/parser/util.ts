export function isDefined<T>(t: T | undefined): t is T {
    return t !== undefined && t !== null;
}
