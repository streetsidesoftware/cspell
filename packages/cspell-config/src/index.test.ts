import * as index from './index';
import { CSpellConfigFileReaderWriter, createReaderWriter } from './index';
import { fixtures } from './test-helpers/fixtures';
import { copyFile, tempPath } from './test-helpers/util';
import { URI } from 'vscode-uri';
import { promises as fs } from 'fs';

describe('index', () => {
    test('index', () => {
        expect(index).toBeDefined();
    });
    test.each`
        value                                  | expected
        ${typeof CSpellConfigFileReaderWriter} | ${'function'}
        ${typeof createReaderWriter}           | ${'function'}
    `('exports', ({ value, expected }) => {
        expect(value).toEqual(expected);
    });
});

describe('cspell-config', () => {
    test.each`
        fixture                                 | addWords
        ${'package/with-value/package.json'}    | ${['apple']}
        ${'package/without-value/package.json'} | ${['apple']}
        ${'cspell.jsonc'}                       | ${['apple', 'cache']}
        ${'cspell.yaml'}                        | ${['apple', 'cache']}
    `('edit config', async ({ fixture, addWords }) => {
        const fixtureFile = fixtures(fixture);
        const tempFile = tempPath(fixture);
        await copyFile(fixtureFile, tempFile);
        const rw = createReaderWriter();
        const uri = URI.file(tempFile).toString();
        const cfg = await rw.readConfig(uri);
        cfg.addWords(addWords);
        await rw.writeConfig(cfg);
        expect(await fs.readFile(tempFile, 'utf-8')).toMatchSnapshot();
    });
});
