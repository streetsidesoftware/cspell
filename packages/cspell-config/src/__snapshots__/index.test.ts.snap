// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cspell-config edit config 1`] = `
"{
  \\"name\\": \\"test-package\\",
  \\"cspell\\": {
    \\"words\\": [
      \\"apple\\",
      \\"banana\\"
    ]
  }
}
"
`;

exports[`cspell-config edit config 2`] = `
"{
  \\"name\\": \\"test-package\\",
  \\"cspell\\": {
    \\"words\\": [
      \\"apple\\"
    ]
  }
}
"
`;

exports[`cspell-config edit config 3`] = `
"{
    // file version
    \\"version\\": \\"0.2\\",
    \\"words\\": [
        \\"angle\\", // with comment
        \\"apple\\",
        // Before \\"before\\"
        \\"before\\",
        \\"cache\\",
        \\"zebra\\" // zebra comment
    ]
}
"
`;

exports[`cspell-config edit config 4`] = `
"version: \\"0.2\\"
words:
  - angle
  - apple
  - before
  - cache
  - zebra
"
`;
